### PR TITLE
GraphQL - Added sort by options to Products GraphQL type

### PR DESCRIPTION
**Pull request by zbigniewkuras**
### Description
Added available sort options and default sort option for product listing to page_info object of Products GraphQL object.

### Fixed Issues (if relevant)
1. magento/graphql-ce#3: Return 'Available sort options' for product listing

### Manual testing scenarios
1. Add following object to query object page_info and run:
`sort_fields {
        default
        options {
          key
          label
        }
      }`
2. Magento should returns object sort_fields contains default and options 

### Contribution checklist
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products.php
Comment (ishakhsuvarov) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products.php]: Please add constructor doc block type hint for `StoreManagerInterface`

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products.php
Comment (zbigniewkuras) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products.php]: Added!

**Review by ishakhsuvarov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Closed and Re Opened PR to trigger Travis CI builds.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products.php
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Products.php]: Please create a separate resolver for `sort_fields`. There are 2 main reasons for that:
1. Performance. In case `sort_fields` is not requested, all this logic will not be executed in case of separate resolver
2. Following single responsibility and object decomposition simplifies maintenance and understanding

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=12) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @zbigniewkuras 
Please sign [Contributor License Agreement](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=12) for this Pull Request so that we could accept this Pull Request. Thank you.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Products.php
Comment (zbigniewkuras) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Products.php]: Roger, i will fix it, thanks for your feedback

File: app/code/Magento/GraphQl/etc/schema.graphqls
Comment (paliarush) [inline app/code/Magento/GraphQl/etc/schema.graphqls]: There should be a **single** new resolver, just for `sort_fields`.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFieldsOptions.php
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFieldsOptions.php]: Please use `\Magento\Catalog\Model\Category\Attribute\Source\Sortby::getAllOptions`. It has two benefits:
- includes position, so no need to hardcode it here again
- result is already properly formatted, no need to do foreach (assuming that `key` is renamed to `value` as suggested in another comment)

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductSearchTest.php
Comment (paliarush) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductSearchTest.php]: Please rename `key` to `value` for consistency with naming in other places in Magento, for example `\Magento\Catalog\Model\Category\Attribute\Source\Sortby::getAllOptions`

File: dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductSearchTest.php
Comment (paliarush) [inline dev/tests/api-functional/testsuite/Magento/GraphQl/Catalog/ProductSearchTest.php]: Please add assertion for `options` values, it is not validated at the moment.

**Review by paliarush — CHANGES_REQUESTED** [review]
@zbigniewkuras thank you for making additional changes, please see new comments. As soon as those are addressed, we can merge the PR.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFieldsOptions.php
Comment (zbigniewkuras) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFieldsOptions.php]: The reason i didn't use this function was the translation function called for labels in ```\Magento\Catalog\Model\Category\Attribute\Source\Sortby::getAllOptions```  casuses Internal server error and every value for label is equal null. I'm not sure exactly, but it looks like a bug.

Comment (paliarush) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFieldsOptions.php]: How about this one `\Magento\Catalog\Model\Config::getAttributeUsedForSortByArray`?

Comment (zbigniewkuras) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFieldsOptions.php]: It's the same, this function also contains translation function for Position option
``` $options = ['position' => __('Position')];```

File: app/code/Magento/GraphQl/etc/schema.graphqls
Comment (paliarush) [inline app/code/Magento/GraphQl/etc/schema.graphqls]: `sort_fields` should be moved to `Products` type, as mentioned in the issue description, because `SearchResultPageInfo` is meant for storing paginating information.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFields.php
Comment (paliarush) [inline app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFields.php]: Looks like an accidental change.

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFieldsOptions.php
Comment (paliarush) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFieldsOptions.php]: The problem is that underlying GraphQL library (webonyx) expects exact match between declared type (string) and the actual value type (Magento\Framework\Phrase), and it does not try to perform type casting. 

I was able to make it work by manually casting Phrase to string before returning the result:
```
        $sortFieldsOptions = $this->sortBy->getAllOptions();
        array_walk(
            $sortFieldsOptions,
            function (&$option) {
                $option['label'] = (string)$option['label'];
            }
        );
```

File: app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFields.php
Comment (zbigniewkuras) [reply app/code/Magento/CatalogGraphQl/Model/Resolver/Category/SortFields.php]: Yes, you are right. I was testing function ```\Magento\Catalog\Model\Category\Attribute\Source\Sortby::getAllOptions()```. I will fix it

File: app/code/Magento/GraphQl/etc/schema.graphqls
Comment (zbigniewkuras) [reply app/code/Magento/GraphQl/etc/schema.graphqls]: Okey

**Comment by misha-kotov** [comment]
Approved :white_check_mark:

**Review by paliarush — APPROVED** [review]

**Comment by paliarush** [comment]
@zbigniewkuras, the PR has been merged to `mainline/2.3-develop`, thanks you for contribution!

**Comment by keharper** [comment]
Added SortFields information in the [Products endpoint](https://devdocs.magento.com/guides/v2.3/graphql/reference/products.html) documentation.

**Pull request merged**
